File: python-resend-examples/examples/broadcasts.py
#!/usr/bin/env python3
"""
Broadcast Clicked Links

Demonstrates retrieving the links clicked in a broadcast, ranked by total
clicks, with cursor-based pagination.

Usage:
    python examples/broadcasts.py

See: https://resend.com/docs/api-reference/broadcasts/list-broadcast-clicked-links
"""

import os
import resend
from dotenv import load_dotenv

load_dotenv()

resend.api_key = os.environ["RESEND_API_KEY"]

# Use an existing broadcast, or create one in the dashboard
broadcast_id = os.environ.get("RESEND_BROADCAST_ID", "your-broadcast-id")

print("=== Broadcast Clicked Links ===\n")

# List broadcasts to see what's available
print("Listing broadcasts...")
broadcasts = resend.Broadcasts.list()
print(f"Found {len(broadcasts.get('data', []))} broadcast(s)")
print()

# List the links clicked in a broadcast, ranked by total clicks
print("Listing clicked links...")
clicked_links = resend.Broadcasts.clicked_links(broadcast_id)
print(f"Found {len(clicked_links.get('data', []))} clicked link(s)")
for link in clicked_links.get("data", []):
    print(f"  - {link['url']} ({link['clicks']} clicks, {link['unique_clicks']} unique)")
print()

# Paginate with limit/after/before - `id` on each row is an opaque cursor
# for that row, not an entity id
print("Listing clicked links with pagination...")
page = resend.Broadcasts.clicked_links(broadcast_id, params={"limit": 10})
print(f"Found {len(page.get('data', []))} clicked link(s), has_more={page.get('has_more')}")

if page.get("has_more") and page.get("data"):
    next_cursor = page["data"][-1]["id"]
    print(f"Fetching next page after cursor {next_cursor}...")
    next_page = resend.Broadcasts.clicked_links(
        broadcast_id, params={"limit": 10, "after": next_cursor}
    )
    print(f"Found {len(next_page.get('data', []))} more clicked link(s)")

File: rust-resend-examples/examples/broadcasts.rs
use resend_rs::list_opts::ListOptions;
use resend_rs::Resend;

#[tokio::main]
async fn main() {
    dotenvy::dotenv().ok();

    let api_key = std::env::var("RESEND_API_KEY").expect("RESEND_API_KEY environment variable is required");
    let broadcast_id = std::env::var("RESEND_BROADCAST_ID").unwrap_or_else(|_| "your-broadcast-id".to_string());

    let resend = Resend::new(&api_key);

    // 1. List broadcasts, to show what's available
    println!("=== Listing Broadcasts ===");
    match resend.broadcasts.list(ListOptions::default()).await {
        Ok(broadcasts) => {
            for broadcast in &broadcasts.data {
                println!("  - {} ({})", broadcast.name, broadcast.id);
            }
        }
        Err(e) => {
            eprintln!("Error listing broadcasts: {}", e);
            std::process::exit(1);
        }
    }

    // 2. List the links clicked in a broadcast, ranked by total clicks
    println!("\n=== Listing Clicked Links ===");
    let first_page = match resend
        .broadcasts
        .clicked_links(&broadcast_id, ListOptions::default().with_limit(2))
        .await
    {
        Ok(links) => {
            println!("Total clicked links on this page: {}", links.data.len());
            for link in &links.data {
                println!(
                    "  - {} (clicks: {}, unique_clicks: {})",
                    link.url, link.clicks, link.unique_clicks
                );
            }
            links
        }
        Err(e) => {
            eprintln!("Error listing clicked links: {}", e);
            std::process::exit(1);
        }
    };

    // 3. Paginate forward if there are more — `id` on each link is an opaque
    //    cursor for that row, not an entity id, and is only meant to be used
    //    with `list_after`/`list_before`.
    if first_page.has_more {
        if let Some(last_link) = first_page.data.last() {
            println!("\n=== Listing Clicked Links (next page) ===");
            let next_page_opts = ListOptions::default().with_limit(2).list_after(&last_link.id);

            match resend.broadcasts.clicked_links(&broadcast_id, next_page_opts).await {
                Ok(links) => {
                    for link in &links.data {
                        println!(
                            "  - {} (clicks: {}, unique_clicks: {})",
                            link.url, link.clicks, link.unique_clicks
                        );
                    }
                }
                Err(e) => eprintln!("Error listing next page of clicked links: {}", e),
            }
        }
    } else {
        println!("\nNo further pages of clicked links.");
    }

    println!("\nDone!");
}
